onnxruntime/core/providers/cpu/tensor/utils.h

```diff
@@ -277,4 +277,115 @@ inline void CopyCpuTensor(const Tensor* src, Tensor* tgt) {
   }
 }
 
+// This provides easy sequential iteration over a subset of a tensor given a span of starts, extents & optionally steps
+template <typename T>
+struct WritableSliceIterator {
+  WritableSliceIterator(Tensor& tensor, gsl::span<const int64_t> starts,
+                        gsl::span<const int64_t> extents, gsl::span<const int64_t> steps)
+      : tensor_(tensor), input_(tensor_.template MutableData<T>()), extents_(extents), skips_(tensor_.Shape(), extents, steps), indices_(extents.size(), 0) {
+    auto& dims = tensor_.Shape().GetDims();
+    Init(dims, starts, steps);
+  }
+
+  // This construct takes a explicit tensor_shape which might be different from the shape defined in input tensor.
+  // The explicit tensor_shape usually has inner most axis flattened. For example, given shape[1,4,4,2], if last axis
+  // does not have padding or slice, then it will be flattened as [1,4,8] for better performance (One inner most copy instead of 4).
+  // Also supports arbitrary positive and negative stepping along individual axes
+  WritableSliceIterator(Tensor& tensor, const TensorShape& tensor_shape, gsl::span<const int64_t> starts,
+                        gsl::span<const int64_t> extents, gsl::span<const int64_t> steps)
+      : tensor_(tensor), input_(tensor_.template MutableData<T>()), extents_(extents), skips_(tensor_shape, extents, steps), indices_(extents.size(), 0) {
+    auto& dims = tensor_shape.GetDims();
+    Init(dims, starts, steps);
+  }
+
+  // Initialize initial skip and inner_extent.
+  void Init(const std::vector<int64_t>& dims, gsl::span<const int64_t> starts,
+            gsl::span<const int64_t> steps) {
+    ORT_ENFORCE(static_cast<ptrdiff_t>(dims.size()) == starts.size() &&
```

Comment:
Append a message to ORT_ENFORCEs for better debuggability.

Reply:
> Append a message to ORT_ENFORCEs for better debuggability.

Added friendlier error message for these conditions separately.